src/IdentityModel.AspNetCore.OAuth2Introspection/Infrastructure/AsyncLazy.cs

```diff
@@ -2,14 +2,41 @@
 // Licensed under the Apache License, Version 2.0. See LICENSE in the project root for license information.
 
 using System;
+using System.Threading;
 using System.Threading.Tasks;
 
 namespace IdentityModel.AspNetCore.OAuth2Introspection.Infrastructure
 {
-    internal class AsyncLazy<T> : Lazy<Task<T>>
+    internal sealed class AsyncLazy<T>
     {
-        public AsyncLazy(Func<Task<T>> taskFactory) :
-            base(() => Task.Factory.StartNew(taskFactory).Unwrap())
-        { }
+        private Lazy<Task<T>> _lazyTaskFactory;
+        private readonly Func<Task<T>> _taskFactory;
+        private readonly object _mutex = new object();
+
+        public AsyncLazy(Func<Task<T>> taskFactory)
+        {
+            _taskFactory = taskFactory;
+            _lazyTaskFactory = InitLazy(_taskFactory);
+        }
+ 
+        public Task<T> GetAsync()
+        {
+            if (!_lazyTaskFactory.IsValueCreated || !_lazyTaskFactory.Value.IsFaulted) return _lazyTaskFactory.Value;
```

Comment:
At the moment if the value is created you will always enter the "mutex" (I'd rename that to guard or something as its not a mutex). If you just check for not faulted you will get the correct behavior

Reply:
If the value is created and the Task is not faulted, it will not enter the lock.

Reply:
Oh sorry - you're right 

might be a bit clearer to write

`if(!(_lazyTaskFactory.IsValueCreated &&_lazyTaskFactory.Value.IsFaulted))`

Reply:
Done. Also added some comments to better explain what is happening.